### PR TITLE
Update to CLDR 40

### PR DESCRIPTION
**Pull request by ehoogerbeets**
- Update data to CLDR 40 and any unit tests that go with it
    - Non-locale updates:
        - Unicode string normalization
        - CType updates for the latest Unicode Character Database
        - Update to latest ISO-15924
            - Adds 7 scripts to the list: Ougr, Pcun, Pelm, Psin, Ranj, Tnsa, Vith
            - Various updates to other scripts
    - Updates for various locales:
        - dates/times
        - duration formats
        - date range formats
        - unit formats
        - list formats
        - number formats
        - clock 12/24 prefs
- Update to Unicode Character Database v14
- CircleCI now checks ilib unit tests on various versions of node: 8, 10, 12, 14, 16

### Checklist

* [x] At least one test case is included for this feature or bug fix.
* [x] `ReleaseNotes` has added or is not needed.
* [x] This PR has passed all of the test cases on `Nodejs` and `Chrome Browser`.
* [x] This PR has passed all of the test cases on `QT/QML`.
* [ ] This is an API breaking change.
* [ ] Requires a major version change.

### Issue Resolved / Feature Added
[//]: # (Describe the issue resolved or feature added by this pull request)

### Resolution
[//]: # (Does the code work as intended?)
[//]: # (What is the impact of this change and *why* was it made?)

### Links
[//]: # (Related issues, references)

### CODE REVIEW

**Comment by gouniLee** [comment]
It's a huge diff. I think it would have been nicer if it was separated into several PRs. I need to check the differences since 38 from webOS perspective so give me enough time to check them all.

File: js/data/locale/ceb/dateformats.json [hunk 1]
```diff
@@ -56,72 +50,250 @@
...
+            },
+            "c30": {
+                "f": "{sy} – {ey}",
+                "l": "{sy} – {ey}",
+                "m": "{sy} – {ey}",
+                "s": "{sy} – {ey}"
+            }
+        }
+    },
+    "islamic": "gregorian",
```
Comment (gouniLee) [inline js/data/locale/ceb/dateformats.json]: many `dateformat.json` files have these part. line from 166. 
It's duplicated and I think enough to have in root.

File: js/data/locale/af/NA/dateformats.json [hunk 2]
```diff
@@ -3,46 +3,40 @@
         "order": "{date} {time}",
         "date": {
             "dmwy": {
-                "f": "EEEE dd MMMM yyyy",
-                "l": "EEE dd MMMM yyyy",
+                "s": "E yyyy-MM-dd",
                 "m": "EE dd MMM yyyy",
-                "s": "E yyyy-MM-dd"
+                "l": "EEE dd MMMM yyyy",
```
Comment (gouniLee) [inline js/data/locale/af/NA/dateformats.json:8]: The reason to have many diff in `dateformats.json` is just  changed the data format order.. by changed "to s" come first

File: js/data/locale/ak/dateformats.json [hunk 3]
```diff
@@ -51,98 +45,276 @@
...
+                "l": "EEE",
+                "f": "EEEE"
             }
         },
         "time": {
             "12": {
-                "h": "h",
-                "m": "mm",
-                "s": "ss",
+                "ahmsz": "h:mm:ss a z",
```
Comment (gouniLee) [inline js/data/locale/ak/dateformats.json:74]: diff is just changed the order..

File: js/data/locale/af/NA/dateformats.json (hunk 2 above)
Comment (ehoogerbeets) [reply js/data/locale/af/NA/dateformats.json:8]: true, we should sort them so that they always come out in the same order

Comment (ehoogerbeets) [reply js/data/locale/af/NA/dateformats.json:8]: or at least always after this

Comment (gouniLee) [reply js/data/locale/af/NA/dateformats.json]: yes. I re-generated dateformats.json and sysres.json files by standalone running gendatefmt2.js file. It saves files changes from 1830 -> 1537 in this PR.
As you said we should sort them to get same order all the time.

Comment (ehoogerbeets) [reply js/data/locale/af/NA/dateformats.json]: I just implemented the sorting but it makes such a huge change that it makes it impossible to see the differences. So, I put in the treeSort branch and when this PR is done, I'll make a new PR from that branch that contains the exact same settings, but sorted. That way, we know nothing is changed except the order.

Comment (gouniLee) [reply js/data/locale/af/NA/dateformats.json]: oh I see. That's good : )

File: tools/cldr/gendatefmts2.js [hunk 4]
```diff
@@ -43,14 +43,137 @@ var hardCodeData = {
...
+        "#{num}s": "#{num} с",
+        "durationShortMinutes": "#{num} д",
+        "#{num}h": "#{num} с",
+        "#{num}d": "#{num} р",
+        "#{num}w": "#{num} ҳ",
+        "durationShortMonths": "#{num} м",
+        "#{num}y": "#{num} с"
+    },
+    "mt": {
+        "1#1 second|#{num} seconds": "one#{num} sekonda|few#{num} sekondi|many#{num}-il sekonda|#{num} sekonda",
```
Comment (gouniLee) [inline tools/cldr/gendatefmts2.js:93]: One of  concern is if we add these parts as hardcode, We may miss the chance to get next cldr updates

File: js/data/locale/fa/AF/unitfmt.json [hunk 5]
```diff
@@ -335,52 +356,59 @@
...
-            "cubic-mile": "#{n} مایل مکعب",
-            "cubic-yard": "#{n} یارد مکعب",
-            "cubic-foot": "#{n} فوت مکعب",
-            "cubic-inch": "#{n} اینچ مکعب",
-            "megaliter": "#{n} مگالیتر",
-            "hectoliter": "#{n} هکتولیتر",
-            "liter": "#{n} لیتر",
-            "deciliter": "#{n} دسی‌لیتر",
-            "centiliter": "#{n} سانتی‌لیتر",
+            "cubic-mile": "#{n}mi³",
```
Comment (gouniLee) [inline js/data/locale/fa/AF/unitfmt.json]: note) Usually iLib prefer to provide their own language translation. but The latest CLDR changed unit translation to english

File: js/data/locale/zh/Hans/HK/unitfmt.json [hunk 6]
```diff
@@ -209,29 +223,30 @@
...
-            "terabit": "#{n}太比特",
-            "gigabyte": "#{n}吉字节",
-            "gigabit": "#{n}吉比特",
-            "megabyte": "#{n}兆字节",
-            "megabit": "#{n}兆比特",
-            "kilobyte": "#{n}千字节",
-            "kilobit": "#{n}千比特",
-            "byte": "#{n}字节",
-            "bit": "#{n}比特",
+            "terabyte": "#{n} TB",
```
Comment (gouniLee) [inline js/data/locale/zh/Hans/HK/unitfmt.json]: note) Usually iLib prefer to provide their own language translation. but The latest CLDR changed unit translation to english

File: js/lib/nfc/Arab.js [hunk 7]
```diff
@@ -1,7 +1,7 @@
 /*
  * Arab.js - include file for normalization data for a particular script
  * 
- * Copyright © 2013 - 2018, JEDLSoft
+ * Copyright © 2013 - 2018, 2020 JEDLSoft
```
Comment (gouniLee) [inline js/lib/nfc/Arab.js]: 2020?

File: js/data/locale/ku/dateformats.json [hunk 8]
```diff
@@ -151,5 +157,172 @@
                 "s": "‏{sy} – {ey}"
             }
         }
-    }
+    },
+    "islamic": "gregorian",
```
Comment (gouniLee) [inline js/data/locale/ku/dateformats.json]: `cop/dateformats.json, ku/dateformats.json, prg/dateformats.json
cu/dateformats.json, vo/dateformats.json` files are still have duplicated contents from root

File: js/data/locale/fa/AF/unitfmt.json (hunk 5 above)
Comment (ehoogerbeets) [reply js/data/locale/fa/AF/unitfmt.json]: I would not second-guess what the CLDR folks are doing. They changed it on purpose. Maybe that is the way it is actually represented in that language/country?

The only time we should override is when CLDR does not yet have translations, but ilib has ones previously provided by LG members. Those are for tg, mt, wo, and zu

File: js/lib/nfc/Arab.js (hunk 7 above)
Comment (ehoogerbeets) [reply js/lib/nfc/Arab.js]: updated the tool to generate the right copyright notice

File: js/data/locale/ku/dateformats.json (hunk 8 above)
Comment (ehoogerbeets) [reply js/data/locale/ku/dateformats.json]: Change the cldr tool to run `mergeAndPrune()` on the date formats before writing them out

File: tools/cldr/gendatefmts2.js (hunk 4 above)
Comment (ehoogerbeets) [reply tools/cldr/gendatefmts2.js:93]: That is true... we will have to be vigilant when we update to CLDR41 to see if CLDR has added those locales yet.

File: js/data/locale/ak/dateformats.json (hunk 3 above)
Comment (ehoogerbeets) [reply js/data/locale/ak/dateformats.json:54]: Order will be addressed in a separate PR once everything is in development and all unit tests pass.

File: js/data/locale/ceb/dateformats.json (hunk 1 above)
Comment (ehoogerbeets) [reply js/data/locale/ceb/dateformats.json]: Ran `mergeAndPrune` on all the date formats data

**Review by gouniLee — APPROVED** [review]
Because huge diff and update, there's no additional newly added test case regarding cldr update. I noticed plural.json update is missing.  it's ok to handle different PR.

**Pull request merged**